From b5f5d4489bd26e3f88c121d82aa960090c909025 Mon Sep 17 00:00:00 2001
From: Devi Priya <quic_devipriy@quicinc.com>
Date: Fri, 5 Sep 2025 12:22:53 +0530
Subject: [PATCH] i2c: qcom-geni: Add i2c_clk_fld table for FW based SE

Add support for i2c_clk_fld table for FW based SEs. Also, check
if the SE is minicore/firmware based and update the counter
values accordingly. Include frequency tolerance check to consider
clock rates approximating 32MHz as valid.

While at it, Update the cycle counter value for minicore SE to meet the
frequency req in standard mode (< 100KHz).

Change-Id: I38f8c86e3f12eef0590237cc7f5c139666a41767
Signed-off-by: Devi Priya <quic_devipriy@quicinc.com>
---
 drivers/i2c/busses/i2c-qcom-geni.c | 45 ++++++++++++++++++++++++------
 drivers/soc/qcom/qcom-geni-se.c    |  2 --
 include/linux/soc/qcom/geni-se.h   |  2 ++
 3 files changed, 39 insertions(+), 10 deletions(-)

diff --git a/drivers/i2c/busses/i2c-qcom-geni.c b/drivers/i2c/busses/i2c-qcom-geni.c
index c60925d6b685..85123066a35d 100644
--- a/drivers/i2c/busses/i2c-qcom-geni.c
+++ b/drivers/i2c/busses/i2c-qcom-geni.c
@@ -77,6 +77,7 @@ enum geni_i2c_err_code {
 #define ABORT_TIMEOUT		HZ
 #define XFER_TIMEOUT		HZ
 #define RST_TIMEOUT		HZ
+#define I2C_32MHZ_TOLERANCE_HZ  320000 /* 1% Tolerance for 32 MHz */
 
 struct geni_i2c_dev {
 	struct geni_se se;
@@ -155,21 +156,49 @@ static const struct geni_i2c_clk_fld geni_i2c_clk_map_19p2mhz[] = {
 };
 
 /* source_clock = 32 MHz */
-static const struct geni_i2c_clk_fld geni_i2c_clk_map_32mhz[] = {
-       {KHZ(100), 12, 9, 10, 25},
-       {KHZ(400), 4,  3,  9, 19},
-       {KHZ(1000), 2, 3,  5, 15},
-       {},
+static const struct geni_i2c_clk_fld geni_i2c_clk_map_32mhz_minicore_se[] = {
+	{KHZ(100), 12, 9, 10, 26},
+	{KHZ(400), 4,  3,  9, 19},
+	{KHZ(1000), 2, 3,  5, 15},
+	{},
+};
+
+static const struct geni_i2c_clk_fld geni_i2c_clk_map_32mhz_firmware_se[] = {
+	{KHZ(100), 12, 9, 10, 25},
+	{KHZ(400), 4,  3,  9, 18},
+	{KHZ(1000), 2, 3,  4, 14},
+	{},
 };
 
 static int geni_i2c_clk_map_idx(struct geni_i2c_dev *gi2c)
 {
 	const struct geni_i2c_clk_fld *itr;
+	struct geni_se *se = &gi2c->se;
+	u32 val;
+	unsigned long clk_rate;
 
-	if (clk_get_rate(gi2c->se.clk) == 32 * HZ_PER_MHZ)
-		itr = geni_i2c_clk_map_32mhz;
-	else
+	/* Check if the SE is mini-core or Firmware based to
+	 * pick the counter values accordingly.
+	 * Add frequency tolerance check to ensure clock rates
+	 * approximating 32MHz (eg:31.9MHz) as valid and pick
+	 * appropriate counter values.
+	 */
+	if (!gi2c->se.clk) {
+		dev_err(gi2c->se.dev, "SE clock is not initialized\n");
+		return -EINVAL;
+	}
+
+	clk_rate = clk_get_rate(gi2c->se.clk);
+	if (abs((long)clk_rate - (long)(32 * HZ_PER_MHZ))
+	    <= I2C_32MHZ_TOLERANCE_HZ) {
+		val = readl_relaxed(se->base + SE_HW_PARAM_2_OFFSET);
+		if (val & SE_GENI_USE_MINICORES_MASK)
+			itr = geni_i2c_clk_map_32mhz_minicore_se;
+		else
+			itr = geni_i2c_clk_map_32mhz_firmware_se;
+	} else {
 		itr = geni_i2c_clk_map_19p2mhz;
+	}
 
 	while (itr->clk_freq_out != 0) {
 		if (itr->clk_freq_out == gi2c->clk_freq_out) {
diff --git a/drivers/soc/qcom/qcom-geni-se.c b/drivers/soc/qcom/qcom-geni-se.c
index 7741c95b0cee..ee7acbdeb933 100644
--- a/drivers/soc/qcom/qcom-geni-se.c
+++ b/drivers/soc/qcom/qcom-geni-se.c
@@ -900,13 +900,11 @@ EXPORT_SYMBOL(geni_icc_disable);
 #define SE_GENI_OUTPUT_CTRL_OFFSET		0x0024
 #define SE_GENI_CGC_CTRL_OFFSET			0x0028
 #define SE_GENI4_IMAGE_REG_OFFSET		0x0100
-#define SE_HW_PARAM_2_OFFSET			0x0E2C
 #define SE_GENI_FW_REVISION_OFFSET		0x1000
 #define SE_GENI_S_FW_REVISION_OFFSET		0x1004
 #define SE_GENI_CFG_RAMn_OFFSET			0x1010
 #define SE_GENI_CLK_CTRL_OFFSET			0x2000
 #define SE_DMA_IF_EN_OFFSET			0x2004
-#define SE_GENI_USE_MINICORES_MASK		BIT(12)
 
 typedef struct __attribute__ ((packed)) se_fw_hdr
 {
diff --git a/include/linux/soc/qcom/geni-se.h b/include/linux/soc/qcom/geni-se.h
index f1e38786beb9..85a11e214164 100644
--- a/include/linux/soc/qcom/geni-se.h
+++ b/include/linux/soc/qcom/geni-se.h
@@ -126,6 +126,8 @@ struct geni_se {
 #define SE_DMA_RX_FSM_RST		0xd58
 #define SE_HW_PARAM_0			0xe24
 #define SE_HW_PARAM_1			0xe28
+#define SE_HW_PARAM_2_OFFSET		0xe2c
+#define SE_GENI_USE_MINICORES_MASK	BIT(12)
 
 /* GENI_FORCE_DEFAULT_REG fields */
 #define FORCE_DEFAULT	BIT(0)
-- 
2.34.1

